modcc/parser.cpp

```diff
@@ -558,6 +561,68 @@ void Parser::parse_parameter_block() {
     return;
 }
 
+void Parser::parse_constant_block() {
+    get_token();
+
+    // assert that the block starts with a curly brace
+    if(token_.type != tok::lbrace) {
+        error(pprintf("CONSTANT block must start with a curly brace {, found '%'", token_.spelling));
+        return;
+    }
+
+    int success = 1;
+    get_token();
+    while(token_.type!=tok::rbrace && token_.type!=tok::eof) {
+        int line = location_.line;
+        std::string name, value;
+
+        // read the constant name
+        if(token_.type != tok::identifier) {
+            success = 0;
+            goto parm_exit;
+        }
+        name = token_.spelling; // save full token
+
+        get_token();
+
+        // look for equality
+        if(token_.type==tok::eq) {
+            get_token(); // consume '='
+            value = value_literal();
+            if(status_ == lexerStatus::error) {
+                success = 0;
+                goto parm_exit;
+            }
+        }
+
+        // get the units
+        if(line==location_.line && token_.type == tok::lparen) {
+            unit_description();
+            if(status_ == lexerStatus::error) {
+                success = 0;
+                goto parm_exit;
+            }
+        }
+
+        constants_map_.insert({name, value});
+    }
+
+    // error if EOF before closing curly brace
+    if(token_.type==tok::eof) {
+        error("CONSTANT block must have closing '}'");
+        goto parm_exit;
+    }
+
+    get_token(); // consume closing brace
+
+    parm_exit:
+    // only write error message if one hasn't already been logged by the lexer
+    if(!success && status_==lexerStatus::happy) {
+        error(pprintf("CONSTANT block unexpected symbol '%s'", token_.spelling));
+    }
+    return;
+}
```

Comment:
The error handling here is a bit convoluted. I think it could be simplified by:
* There are only three points, at lines 569, 582, and 612, where a non-lexer error condition is handled. Each has a different message, so there's nothing really to be gained by jumping on to `parm_exit` for the last two. I suggest using `error(...)` and return in each of the three instances.
* When there is a lexer error, no further error will be logged, so one can just return directly (lines 593, 602).
* This in turn means there is no need form the `parm_exit:` code at all, I think.

Reply:
I admit I don't understand the whole picture when it comes to the error handling code in the parser. I'll implement your suggestions, thanks!